### DISCUSSION TITLE
Type Inference for `decodeEventLog` not inferring the correct value

### DISCUSSION THREAD

**Discussion by mulf0** [wevm/viem#380] [Question] [answered]
I have this bit of code spitting out all of the logs for a given ERC20 contract. Based on the docs for `decodeEventLog`, I should get a return type that looks like this:

```ts
/**
 *  {
 *    eventName: 'Transfer',
 *    args: {
 *      from: '0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266',
 *      to: '0x70997970c51812dc3a010c7d01b50e0d17dc79c8'
 *      value: 1n
 *    }
 *  }
 */
```

However, what I actually get is 
![Screenshot 2023-04-16 at 2 37 27 PM](https://user-images.githubusercontent.com/63870279/232337802-e12885fa-6516-43d6-8552-628cd5628e55.png)

This causes the required type assertion in the included code. After debugging a bit I could not get anything other than the empty object as the inferred type. Any help would be wonderful. Package version `0.3.2`

```ts
import { createPublicClient, decodeEventLog, http, parseAbiItem } from "viem";
import { mainnet } from "viem/chains";
import { erc20ABI } from "@wagmi/core";

type TransferLogArgs = {
  from: string;
  to: string;
  value: BigInt;
};

const transport = http(
  "https://eth-mainnet.g.alchemy.com/v2/my-correct-api-key"
);

const client = createPublicClient({
  chain: mainnet,
  transport,
});

async function main() {
  const BLOCK_INTERVAL = 250;
  let startBlock = 15390081;


  setInterval(async () => {
    console.log({ startBlock });
    const changes = await client.getLogs({
      address: "0xb24cd494faE4C180A89975F1328Eab2a7D5d8f11",
      event: parseAbiItem(
        "event Transfer(address indexed from, address indexed to, uint256 value)"
      ),
      fromBlock: BigInt(startBlock),
      toBlock: BigInt(startBlock + BLOCK_INTERVAL),
    });

    startBlock += BLOCK_INTERVAL;

    for (const change of changes) {
      const log = decodeEventLog({
        abi: erc20ABI,
        data: change.data,
        topics: change.topics,
        eventName: change.eventName,
      });

      if (typeof log.args === "undefined") {
        return new Error("incorrect data");
      }
  
      const info = log.args as TransferLogArgs;

      console.log(info);
    }
  }, 1500);
}

main().catch((err) => {
  console.log(err);
  process.exit(1);
});

```

**Comment by jxom** [comment]
Looks like a type bug. We can have a look into it! For the meantime, you may want to cast your topics to `[Hex, Hex, Hex]` (the expected number of topics for the log) to get the type inference to work. See line 36:

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQBQK4BGANsMgMIloB2MANHACZrIRMCiAbtTADIQDm9ABYwYYemAxQAzmgCCBYAElsIegAk0ADzgBfOADMoEEHABEHYGhBmA3AFgAUKEixEcEBmBUqaePqMTc0trAHpkIS8qaTsnF2h4JDQoZAAmAAY5ACElPUNjUzMAAQB3DH4QYHDoNFjHJxZo+BgoDGjXeABeOBExAAonOHNesGkALlDQvyEAWk9vXxgAOn4ljCII6wBPJZYQUI5U0JAtmZYodGQYGYwwYBmAazQtsycASgd6x0bpeGRKGhwbqoTDYfDEUgUKw0PoIQYoSLeMYeKKLWjwlptaQddGOXQfJxODDSLZUZCGPBkmDACBUFHePpvRDwn7wLI8ADyZAA0gB9JQAOQAKmwAEoANTkPCBcFSAFZ0p8hkQ-HBflIYFkiBBkA8ZQBGOUAZgAnOl0gAOfWfQmOIayGBKGjJDjrPrE0nkxlAgB8zLtQxQtOkEBVS21-FhapgGq1Or1+KVgdZCLa-DQ0hlGDKwD+AOW6d4AmksPhgbgGAYDHQ0nG5nSWgIqQALMgGM2Tc2DBg2M2yPqLZkLSaTQB2OUAMX1RtSFrYGCbGFHABE5QwLQZ9fqzLjy0M0FwaMjJDJ5IoVNYBgG9+YD9w4ELWtEDMl3VWa5nvEwtGgGPkTPQlbVhmn5UN+v5wDAED0Hg3gwPKABscCukQeBoG8rzXoGby7nugQgHGurIlkwD8E6MB9OqsCEQ8OFloGUE0cRpHkZRMbUdqupwAA1HA7JcnygoihKUp0Vhia2uWVGapxercd0-E8vywpipKPA2lhBjQHAfQphEaZoHAEAGKmVDptITJwlhQwphGMpMCw7CHkWkZWTeQwLsAyLJGkmQ5LhN4MFgGDIvpZloEsQUxgFe5QXcyB1mF6ZLHFpCZsScAANqaFoGjaHlWgALoxeWd40AKGAgGgoWROFSxlTAFVVSVegEl8N7ACZfQwFsYBoMZcARmsUD8JmnTjeYlJMAY3i-mYln0eW6AwHgUB0r4JRwGwFzQH0ZjeOclzwFFGDzUm5a6PCi1Bk0cDeFpMpDVIo2SXuPyhhFEZ9PdEBtRd8K6PQhrmm1l1fPMVCMrsWARH0fTJFATKdH6bk3SGYZfQjf1wGAxjICB9VaLmfT6qDbVOEAA

**Reply by jxom** [reply]
Created a tracking issue: #381

**Comment by fubhy** [comment]
@mhmulford0 Type issue with the second ("dynamic") call aside, why are you doing that second `decodeEventLog` in the first place for this case? `.getLogs()` already does the decoding internally and returns `changes[].args` typed correctly.

**Reply by fubhy** [reply, marked as answer]
This would work just fine:

```typescript
import { createPublicClient, http, parseAbiItem } from 'viem'
import { mainnet } from 'viem/chains'

const transport = http(
  'https://eth-mainnet.g.alchemy.com/v2/my-correct-api-key',
)

const client = createPublicClient({
  chain: mainnet,
  transport,
})

const BLOCK_INTERVAL = 250
let startBlock = 15390081

setInterval(async () => {
  console.log({ startBlock })
  const event = parseAbiItem(
    'event Transfer(address indexed from, address indexed to, uint256 value)',
  )
  const changes = await client.getLogs({
    address: '0xb24cd494faE4C180A89975F1328Eab2a7D5d8f11',
    event: event,
    fromBlock: BigInt(startBlock),
    toBlock: BigInt(startBlock + BLOCK_INTERVAL),
  })

  startBlock += BLOCK_INTERVAL

  for (const change of changes) {
    console.log(change.args)
  }
}, 1500)
```

**Reply by mulf0** [reply]
> @mhmulford0 Type issue with the second ("dynamic") call aside, why are you doing that second `decodeEventLog` in the first place for this case? `.getLogs()` already does the decoding internally and returns `changes[].args` typed correctly.

It appears I made an incorrect assumption when reading the docs! I presumed that `getLogs` retrieved them and `decodeEventLog` made them usable. The above code makes a lot more sense, thank you.
